File: sdk-python/arcflow/context.py
"""Agent context and tool-loop configuration (Phase 2-Pro / RCS v0.6)."""

from __future__ import annotations

import json
from enum import Enum

from arcflow.exceptions import WorkflowConfigurationError


class PriorStepsMode(str, Enum):
    """Which prior step outputs to include in agent context."""

    ALL = "all"
    LAST = "last"
    NONE = "none"


class ContextPolicy:
    """Controls how prior steps and run input are assembled into agent prompts."""

    def __init__(
        self,
        *,
        include_prior_steps: PriorStepsMode | str = PriorStepsMode.ALL,
        include_run_input: bool = True,
        max_prior_step_chars: int = 4096,
    ) -> None:
        if isinstance(include_prior_steps, str):
            try:
                include_prior_steps = PriorStepsMode(include_prior_steps.lower())
            except ValueError as exc:
                raise WorkflowConfigurationError(
                    "[ArcFlow] include_prior_steps must be 'all', 'last', or 'none'."
                ) from exc
        if max_prior_step_chars < 256:
            raise WorkflowConfigurationError(
                "[ArcFlow] max_prior_step_chars must be at least 256."
            )
        self.include_prior_steps = include_prior_steps
        self.include_run_input = include_run_input
        self.max_prior_step_chars = max_prior_step_chars

    def binding_json(self) -> str:
        payload = {
            "include_prior_steps": self.include_prior_steps.value,
            "include_run_input": self.include_run_input,
            "max_prior_step_chars": self.max_prior_step_chars,
        }
        return json.dumps(payload)


class ToolExecutionConfig:
    """Bounds LLM function-calling tool loops on an agent."""

    def __init__(
        self,
        *,
        mode: str = "llm_select",
        max_iterations: int = 5,
    ) -> None:
        if mode not in ("legacy_eager", "llm_select"):
            raise WorkflowConfigurationError(
                "[ArcFlow] tool_execution.mode must be 'legacy_eager' or 'llm_select'."
            )
        if max_iterations < 1 or max_iterations > 20:
            raise WorkflowConfigurationError(
                "[ArcFlow] tool_execution.max_iterations must be between 1 and 20."
            )
        self.mode = mode
        self.max_iterations = max_iterations

    def binding_json(self) -> str:
        return json.dumps({"mode": self.mode, "max_iterations": self.max_iterations})

File: packages/arcflow-static/src/context.ts
import { StaticConfigurationError } from "./errors.js";

export type PriorStepsMode = "all" | "last" | "none";

export class ContextPolicy {
  readonly includePriorSteps: PriorStepsMode;
  readonly includeRunInput: boolean;
  readonly maxPriorStepChars: number;

  constructor(options: {
    includePriorSteps?: PriorStepsMode;
    includeRunInput?: boolean;
    maxPriorStepChars?: number;
  } = {}) {
    this.includePriorSteps = options.includePriorSteps ?? "all";
    if (!["all", "last", "none"].includes(this.includePriorSteps)) {
      throw new StaticConfigurationError(
        "[ArcFlow] includePriorSteps must be 'all', 'last', or 'none'.",
      );
    }
    this.includeRunInput = options.includeRunInput ?? true;
    this.maxPriorStepChars = options.maxPriorStepChars ?? 4096;
    if (this.maxPriorStepChars < 256) {
      throw new StaticConfigurationError("[ArcFlow] maxPriorStepChars must be at least 256.");
    }
  }

  toJson(): Record<string, unknown> {
    return {
      include_prior_steps: this.includePriorSteps,
      include_run_input: this.includeRunInput,
      max_prior_step_chars: this.maxPriorStepChars,
    };
  }
}

export class ToolExecutionConfig {
  readonly mode: "legacy_eager" | "llm_select";
  readonly maxIterations: number;

  constructor(options: { mode?: "legacy_eager" | "llm_select"; maxIterations?: number } = {}) {
    this.mode = options.mode ?? "llm_select";
    if (this.mode !== "legacy_eager" && this.mode !== "llm_select") {
      throw new StaticConfigurationError(
        "[ArcFlow] tool_execution.mode must be 'legacy_eager' or 'llm_select'.",
      );
    }
    this.maxIterations = options.maxIterations ?? 5;
    if (this.maxIterations < 1 || this.maxIterations > 20) {
      throw new StaticConfigurationError(
        "[ArcFlow] maxIterations must be between 1 and 20.",
      );
    }
  }

  toJson(): Record<string, unknown> {
    return { mode: this.mode, max_iterations: this.maxIterations };
  }
}
